### PR TITLE
Move the Reader from / to /read

### PR DESCRIPTION
**Pull request by andrewserong**
This PR moves the reader from `/` to `/read` and adds a `loggedInMiddleware` function to handle requests to `/`. That function redirects users from `/` to `/read` and should be a suitable place to (at a later stage) run an A/B test on the redirect path, ensuring that test group assignment only occurs on requests to `/`.

This PR was originally going to incorporate an A/B test to redirect `/` to `/home`, however as was suggested in the comments, this will be done in a later PR.

#### Changes proposed in this Pull Request

* Move reader from `/` to `/read`
* Add `loggedInMiddleware` function to redirect from `/` to `/read`
* Update the stats path for the Reader to `/read`

##### Test links

* Go to `http://calypso.localhost:3000/` while logged in, and you should be redirected to`/read`
* From another section (e.g. the Stats page for a site) use the keyboard shortcut `g` followed by `r` and you should be redirected to `/read`
* Clicking on the `/read` link in the master bar (the button next to My Sites) should go directly to `/read`

##### Test Tracks events

* In your browser console set `localStorage.setItem( 'debug', 'calypso:analytics:tracks*' );`
* Navigate to `http://calypso.localhost:3000/`
* This should redirect you to `/read` and in the console, you should see the event `calypso_reader_following_loaded` with the parameter `ui_algo` of `following`, along with many other reader events. In each of these the `ui_algo` parameter should be `following`.

image

##### Try to find areas where there are outstanding links to `/` that expect the Reader to be present

* Have I missed any links in this PR?

Note: in any Calypsoify areas of Jetpack sites, the Reader masterbar will still link through to `/`. Updating Jetpack will be a separate task. To reach an example of this, go to a site with a Business plan and installed plugins, and click on Tools > Plugins from the sidebar. This will open in a new window, where the Reader link in the masterbar will still go to `/`.

### CODE REVIEW

**Comment by andrewserong** [comment]
Looks like this has added some code owners 👋 — this is just a WIP at the moment. More context: paObgF-Hn-p2

**Comment by blowery** [comment]
Hi @andrewserong! This is going to be a fun change. :) Have you considered breaking this up into two changes:

1) Move Reader to /read, not under an AB Test and start redirecting / to /read so no one ends up staying on /
1a) Shake out all the bugs that come from that change, like updating the masterbar on external sites, finding places that link to "the reader" at / (emails?), that sort of thing
2) Add the AB Test that pushes some percentage of folks to /home instead of /read

That way you're not trying to address the bugs that will crop up from 1 while also trying to run an ab test.

**Comment by bluefuton** [comment]
> Looks like this has added some code owners

For WIP code, you can avoid the magical notification of code owners by creating a draft PR instead:

https://github.blog/2019-02-14-introducing-draft-pull-requests/

> Have you considered breaking this up into two changes?

^ I would echo @blowery's thoughts here. Let me know if you need any help with moving to `/read`.

**Comment by andrewserong** [comment]
Thanks @blowery, that's a great idea! I'll do this as two separate issues and update this PR to address the _first_ issue of moving the Reader to `/read` (not under an A/B test).

And thanks @bluefuton I totally forgot that draft PR's avoid pinging code owners, but in this case I'm glad I accidentally pinged everyone :) I'll let you know how I go with moving to `/read`.

File: client/reader/index.js
Comment (jsnajdr) [inline client/reader/index.js]: These redirects should now redirect to `/read`. If I understand it correctly, paths like `/read/following` are invalid or legacy. The current code might redirect Reader paths to a path that renders `/home`.

File: client/root/index.js
Comment (jsnajdr) [inline client/root/index.js]: Does this redirect handler really need to be a section? Could it instead be somewhere around [this code](https://github.com/Automattic/wp-calypso/blob/master/client/boot/common.js#L117-L130) which already handles special `/` redirects for desktop app or devdocs? (in logged out mode)

**Comment by jsnajdr** [comment]
There's one optimization that this PR currently breaks: if the user initiallly navigates to `/`, the server knows that it's Reader and will send the JS and CSS chunks for Reader in the initial _index.html_ document. This will stop happening now: server will send only the entrypoint JS and CSS, the client-side code will need to figure out that Reader is requested, and only after that will issue the download requests.

File: client/reader/index.js
Comment (andrewserong) [reply client/reader/index.js]: Thanks, I've just updated these, too.

File: client/root/index.js
Comment (andrewserong) [reply client/root/index.js]: Great point — no, it really doesn't have to be a separate section.

I think I was overthinking it, and imagining a future state where we're A/B testing redirecting to the customer home, and so mightn't want to have the server send the user the Reader chunks. But like you say below, that breaks an existing optimization, and your suggestion is much cleaner. I've updated this to remove the `root` section, and create a `loggedInMiddleware` for handling the redirect.

**Comment by andrewserong** [comment]
Thanks for taking a look at this one @jsnajdr! That's really helpful feedback.

I've moved the redirect handler to a `loggedInMiddleware` function, to go with a similar pattern to the existing `loggedOutMiddleware` function. The redirect now happens within an anonymous function — I've tried to set it up in a way where it'd be easy to inject an A/B test into the `redirectPath` so that test assignment would only happen when a user goes to `/`.

Also, I've restored the `/` path in the section definition for the reader so that the Reader chunks are still sent server-side for requests to `/`. When we get to running an A/B test on the redirect, I guess this means that the Reader chunks will be loaded regardless of where we're sending the user — but in this case, I think it makes sense to keep the control group in an optimized flow, and not worry as much about ensuring that the test variation doesn't load extra chunks that it doesn't need.

What do you think?

**Comment by jsnajdr** [comment]
> Also, I've restored the `/` path in the section definition for the reader so that the Reader chunks are still sent server-side for requests to `/`. When we get to running an A/B test on the redirect, I guess this means that the Reader chunks will be loaded regardless of where we're sending the user

Yes, it's a bit hacky, but it should work well for the purposes of the A/B test. Server-side, the `/` path is used only to pre-load the chunks, i.e., a passive optimization that doesn't break anything if it sends the wrong chunks.

It would be dangerous if the Reader section was server-side rendered.

**Comment by p-jackson** [comment]
If you browse to the reader on iOS there's a banner at the bottom inviting you to open the page in the app. The code that generates the deep link to the reader needs to be aware that `/` doesn't refer to the reader anymore. In fact it doesn't refer to anywhere anymore, I don't think it's possible for the `<AppBanner>` component to appear on the root page anymore.

Maybe the `'returns a valid Reader URI for the root path'` test should be changed to `'returns a valid URI for the root of the reader section'`

**Comment by p-jackson** [comment]
wp-desktop has a magic way of navigating to the reader using IPC.
https://github.com/Automattic/wp-calypso/blob/ef2ef7d095ea63ec03dba32d474047ffbdafa2bf/client/lib/desktop/index.js#L121

This should be changed to `/read`

**Comment by p-jackson** [comment]
Looks like this test is no longer true of the `pathToSection` helper
https://github.com/Automattic/wp-calypso/blob/38ecc7ec8871339b442354da8b988b458df2e58b/client/lib/path-to-section/test/index.js#L14

This is more of a tidy up. This helper function is used in the inline help, which is already suggesting the correct content on the `/read` page.

**Comment by p-jackson** [comment]
This one's an interesting case. After being invited as a `viewer` or `follower` it looks like this code intentionally redirects to the reader.
https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/my-sites/invites/utils.js#L155

I don't know what a `viewer` is, but after inviting a user as a follower it feels like going to the reader is the right thing.

**Comment by p-jackson** [comment]
On this page `/following/manage`, clicking the back button should go to `/read`
Screen Shot on 2019-11-22 at 10:28:43

https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/reader/following-manage/index.jsx#L207

**Comment by p-jackson** [comment]
The "Followed Sites" link in reader sidebar should point to `/read` and should have a blue background when you're on the default reader page.
I think they can both be fixed here:
https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/reader/sidebar/index.jsx#L151

**Comment by p-jackson** [comment]
At the bottom of full post pages on reader (e.g. `/read/feeds/2283439/posts/2482410524`) there's "MORE ON WORDPRESS.COM" link. That should go to `/read`.

https://github.com/Automattic/wp-calypso/blob/5eb5df59d970e69d94726bebecfde46b67cc1e1c/client/blocks/reader-full-post/index.jsx#L296

**Comment by p-jackson** [comment]
The user hasn't liked anything yet, the `/activities/likes` page is empty with a link "Back to Following". The link should be `/read`.

https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/reader/liked-stream/empty.jsx#L37

**Review by p-jackson — DISMISSED** [review]
<<phew>> I think that's it! Apologies for the notification spam.

The bones of this change are good to go, I've just found a few places where `/` has been used to mean "the reader". Other instances of `/` I found are used for "just go somewhere valid" and "go somewhere after logging in", and I think both of those cases should go through the AB test logic.

**Review by jsnajdr — APPROVED** [review]
Nice work! I didn't verify if every single `/` link was converted to `/read`, just the framework-ish parts of the patch -- and these are good :+1:

**Comment by andrewserong** [comment]
Thanks so much for the review and confirming @jsnajdr — I'll fix up a few more of these links that Phil discovered before getting another Zelda review and then merging.

**Comment by andrewserong** [comment]
> If you browse to the reader on iOS there's a banner at the bottom inviting you to open the page in the app. The code that generates the deep link to the reader needs to be aware that `/` doesn't refer to the reader anymore. In fact it doesn't refer to anywhere anymore, I don't think it's possible for the `<AppBanner>` component to appear on the root page anymore.
> 
> Maybe the `'returns a valid Reader URI for the root path'` test should be changed to `'returns a valid URI for the root of the reader section'`.
> https://github.com/Automattic/wp-calypso/blob/0f9a45a2e417db2bf15e3aced401ef8de623e237/client/blocks/app-banner/test/app-banner.jsx#L33

Thanks for finding this one @p-jackson! Looks like AppBanner gets rendered based on the _section_ that a user is on, rather than the path, so I don't think the function needs changing, but I will update the test to refer to the new URL.

**Comment by andrewserong** [comment]
> Looks like this test is no longer true of the `pathToSection` helper
> https://github.com/Automattic/wp-calypso/blob/38ecc7ec8871339b442354da8b988b458df2e58b/client/lib/path-to-section/test/index.js#L14
> 
> This is more of a tidy up. This helper function is used in the inline help, which is already suggesting the correct content on the `/read` page.

I'm thinking of leaving this test as is (I've renamed it slightly) so that it covers that the _default_ root section is still the Reader. This test ensures that somewhere in `sections.js` the root is still mapped to a section, and that that section is the Reader, so I think it'd be good to keep this around, and update it if and when we eventually want to shift the default / fallback path away from the Reader.

**Comment by andrewserong** [comment]
Thanks again for finding all those links @p-jackson — I've updated the PR to address them all, so it's now ready for another review!

**Comment by bluefuton** [comment]
This is all looking good folks - very thorough! Is there a corresponding patch yet for updating Reader links to `/` in the PHP codebase (emails, notifications, subscriptions admin, etc)?

**Comment by andrewserong** [comment]
Thanks @bluefuton!

> This is all looking good folks - very thorough! Is there a corresponding patch yet for updating Reader links to / in the PHP codebase (emails, notifications, subscriptions admin, etc)?

Not yet, my plan was for us to roll out this front-end update first, and then work on the backend changes separately.

This week we've just had a switch in priorities, so I'll hopefully pick up this Reader work again next week.

**Comment by ramonjd** [comment]
Looks like all the change requests are in for this one. 

I've rebased and given it a whirl.

The only thing I noticed was that the `ui_algo` tracks property is `unknown` when navigating to the reader site using keys `g` and `r`, but since these [key combos aren't turn on for production at all](https://github.com/Automattic/wp-calypso/blob/4b5f048bbb81d6a7b7ec1521bf558870c7ac69a9/client/boot/common.js#L348), then I guess it's fine :) 

Ok to launch in your opinion @andrewserong?

**Comment by andrewserong** [comment]
Thanks for taking it for a spin @ramonjd — I think it should be good to launch, and we can do further updates in smaller PRs to prevent this one from continuing to grow if we find other instances of the existing `/read` urls.

**Comment by andrewserong** [comment]
@ramonjd I've fixed up the `ui_algo` issue in https://github.com/Automattic/wp-calypso/pull/37547/commits/074cc81f66aa46bd59e9f4bf1af5447de1f31458 — do you mind taking another quick look when you get a chance? If it all looks good, I'm happy to merge this one in after lunch.

**Comment by adamziel** [comment]
YES! That's some amazing work right there, thank you so much for getting this across the line @andrewserong

**Pull request merged**
